### PR TITLE
Code suggestions for #1210

### PR DESCRIPTION
**Pull request by hyrodium**
# Overview
This PR adds suggestions for #1210.

* Add support for `@SArray rand(rng, 3)`
* Simple output for `@macroexpand @SArray rand(rng, Float32, 4)`
* Better error handling for `@SArray rand(rng, Float32, Float32)`
* Add support for `randn` and `randexp`

# Details
## Add support for `@SArray rand(rng, 3)`

**On the original PR #1210**
```julia
julia> using Random, StaticArrays

julia> rng = Random.MersenneTwister(42)
MersenneTwister(42)

julia> @SArray rand(rng, 3)
ERROR: MethodError: Cannot `convert` an object of type Vector{Float64} to an object of type Int64
[...]
```

**On this PR**
```julia
julia> using Random, StaticArrays

julia> rng = Random.MersenneTwister(42)
MersenneTwister(42)

julia> @SArray rand(rng, 3)
3-element SVector{3, Float64} with indices SOneTo(3):
 0.5331830160438613
 0.4540291355871424
 0.017686826714964354
```

## Simple output for `@macroexpand @SArray rand(rng, Float32, 4)`

**On the original PR #1210**
```julia
julia> @SArray rand(rng, Float32, 4)
4-element SVector{4, Float32} with indices SOneTo(4):
 0.9603034
 0.76554394
 0.46937835
 0.2332462

julia> @macroexpand @SArray rand(rng, Float32, 4)
quote
    #= /home/hyrodium/.julia/dev/StaticArrays/src/SArray.jl:214 =#
    if rng isa StaticArrays.DataType
        #= /home/hyrodium/.julia/dev/StaticArrays/src/SArray.jl:215 =#
        StaticArrays.rand((SArray){(Tuple){Float32, 4}, rng})
    elseif #= /home/hyrodium/.julia/dev/StaticArrays/src/SArray.jl:216 =# rng isa StaticArrays.Integer
        #= /home/hyrodium/.julia/dev/StaticArrays/src/SArray.jl:217 =#
        StaticArrays.rand((SArray){(Tuple){rng, Float32, 4}})
    elseif #= /home/hyrodium/.julia/dev/StaticArrays/src/SArray.jl:218 =# rng isa (StaticArrays.Random).AbstractRNG
        #= /home/hyrodium/.julia/dev/StaticArrays/src/SArray.jl:220 =#
        (StaticArrays.StaticArrays)._rand(rng, Float32, StaticArrays.Size(4), (SArray){StaticArrays.Tuple{4}, (StaticArrays.Random).gentype(Float32)})
    else
        #= /home/hyrodium/.julia/dev/StaticArrays/src/SArray.jl:231 =#
        (StaticArrays.StaticArrays)._rand((StaticArrays.Random).GLOBAL_RNG, rng, StaticArrays.Size(Float32, 4), (SArray){StaticArrays.Tuple{Float32, 4}, (StaticArrays.Random).gentype(rng)})
    end
end
```

**On this PR**
```julia
julia> @SArray rand(rng, Float32, 4)
4-element SVector{4, Float32} with indices SOneTo(4):
 0.22553396
 0.18603194
 0.16922617
 0.9104786

julia> @macroexpand @SArray rand(rng, Float32, 4)
:(StaticArrays._rand_with_Val(SArray, rng, Float32, StaticArrays._int2val(rng), StaticArrays._int2val(Float32), StaticArrays.Val(StaticArrays.tuple(4))))
```

## Add support for `randn` and `randexp`
**On the original PR #1210**
```julia
julia> @SArray randn(rng, 5)
ERROR: TypeError: in Tuple, in parameter, expected Type, got a value of type MersenneTwister
Stacktrace:
[...]

julia> @SArray randexp(rng, 5)
ERROR: TypeError: in Tuple, in parameter, expected Type, got a value of type MersenneTwister
Stacktrace:
[...]
```

**On this PR**
```julia
julia> @SArray randn(rng, 5)
5-element SVector{5, Float64} with indices SOneTo(5):
  0.18702790710363
  0.5181487878771377
  1.4913791170403063
  0.3675627461748204
 -0.8862052960481365

julia> @SArray randexp(rng, 5)
5-element SVector{5, Float64} with indices SOneTo(5):
 1.11449101941971
 1.473948306820646
 0.06666365902783734
 0.39638956977933426
 0.5090311768294857
```

# Note
* I added more test cases.
* I also updated functions for `@SVector` and `SMatrox`.
* The function `_rand_with_Val` is based on the comment (https://github.com/JuliaArrays/StaticArrays.jl/pull/1210#discussion_r1411229495).

### CODE REVIEW

File: src/SArray.jl [hunk 1]
```diff
@@ -239,6 +252,69 @@ function static_array_gen(::Type{SA}, @nospecialize(ex), mod::Module) where {SA}
...
+            end
+        elseif f === :randn || f === :randexp
+            _f = Symbol(:_, f)
+            if length(ex.args) == 1
+                # No support for `@SArray randn()` etc.
+                error("@$SA got bad expression: $(ex)")
+            elseif _isnonnegvec(ex.args[2:end])
+                # for calls like `randn(dims...)`
+                return :($f($SA{$Tuple{$(escall(ex.args[2:end])...)}}))
+            elseif _isnonnegvec(ex.args[3:end])
```
Comment (mateuszbaran) [inline src/SArray.jl:280]:
```suggestion
            elseif length(ex.args) >= 3 && _isnonnegvec(ex.args[3:end])
```
To avoid indexing out of range.

**Review by mateuszbaran — COMMENTED** [review]
I think this is a good idea, thanks for proposing that. Could you add some tests for the new functionality and safeguards for `ex.args` indexing (like the one I've proposed)?

File: src/SArray.jl (hunk 1 above)
Comment (hyrodium) [reply src/SArray.jl:280]: If `length(ex.args) < 3`, then `3:end` will be empty, and the function `_isnonnegvec` returns `false` for empty vectors.

```julia
julia> import StaticArrays._isnonnegvec

julia> args = [1,2,3]
3-element Vector{Int64}:
 1
 2
 3

julia> _isnonnegvec(args[5:end])
false
```

**Comment by hyrodium** [comment]
Should I update `@SVector` and `@SMatrix` too? (x-ref: https://github.com/JuliaArrays/StaticArrays.jl/issues/1214)

File: src/SArray.jl (hunk 1 above)
Comment (mateuszbaran) [reply src/SArray.jl:280]: Right, I forgot about that feature of range indexing.

**Comment by mateuszbaran** [comment]
Yes, I think it would be good to update these two macros too.

**Comment by hyrodium** [comment]
Hmm, I'm thinking of reusing `static_array_gen` in `@SVector` macro etc., but that may takes a while...
Replacing `static_vector_gen` in `@SVector` definition with

```julia
macro SVector(ex)
    ex = static_array_gen(SArray, ex, __module__)
    :(SVector($ex))
end
```

is almost fine, but some tests fail..

**Comment by hyrodium** [comment]
I have updated `static_vector_gen`. I will update `static_matrix_gen` and tests tomorrow.

**Comment by hyrodium** [comment]
Oooh, I didn't know that `n=3; @SVector rand(n)` is supported. 

```julia
julia> n=3; @SVector rand(n)  # current master branch
3-element SVector{3, Float64} with indices SOneTo(3):
 0.41421419554720196
 0.8417028007187902
 0.05502353350873945
```

I assumed all of the array sizes must be integers but not variables in this PR.

```julia
julia> n=3; @SVector rand(n)  # on this PR
ERROR: LoadError: @SVector got bad expression: rand(n)
Stacktrace:
[...]
```

I noticed this behavior when testing `@SMatrix`...

https://github.com/JuliaArrays/StaticArrays.jl/blob/67d9c36c7190008056f1dd0fefd31eb33c444d3e/test/eigen.jl#L283-L284

**Comment by mateuszbaran** [comment]
Yes, that's why some macro variants generate `if` statements.

**Comment by hyrodium** [comment]
I'll update this PR this weekend:muscle:

**Comment by hyrodium** [comment]
Sorry for the late. I think this is ready to review after CI results turn green!
I have also updated the PR description.

**Comment by mateuszbaran** [comment]
Thanks! I will check this PR.

**Review by mateuszbaran — APPROVED** [review]
It looks great, thanks!

**Comment by hyrodium** [comment]
Thanks for the quick review! :smile:

**Pull request merged**